Client/Frontend/Browser/Search/SearchSuggestionDataSource.swift

```diff
@@ -0,0 +1,180 @@
+
+    if quickSearchEngines.count == 1 {
+      return !isDefaultEngineQuickEngine
+    }
+
+    return quickSearchEngines.count > 1
+  }
+  
+  var availableSections: [SearchListSection] {
+    var sections = [SearchListSection]()
+    sections.append(.quickBar)
+
+    if !tabType.isPrivate &&
+        searchEngines?.shouldShowSearchSuggestionsOptIn == true {
+      sections.append(.searchSuggestionsOptIn)
+    }
+
+    if !tabType.isPrivate &&
+        searchEngines?.shouldShowSearchSuggestions == true {
+      sections.append(.searchSuggestions)
+    }
+    sections.append(.findInPage)
+    sections.append(.openTabsAndHistoryAndBookmarks)
+    return sections
+  }
+  
+  var braveSearchPromotionAvailable: Bool {
+    guard Preferences.Review.launchCount.value > 1,
+          searchEngines?.defaultEngine().shortName != OpenSearchEngine.EngineNames.brave,
+          let braveSearchPromotionLaunchDate = Preferences.BraveSearch.braveSearchPromotionLaunchDate.value,
+          Preferences.BraveSearch.braveSearchPromotionCompletionState.value != BraveSearchPromotionState.dismissed.rawValue,
+          Preferences.BraveSearch.braveSearchPromotionCompletionState.value != BraveSearchPromotionState.maybeLaterSameSession.rawValue else {
+      return false
+    }
+    
+    let rightNow = Date()
+    let nextShowDate = braveSearchPromotionLaunchDate.addingTimeInterval(
+      AppConstants.buildChannel.isPublic ? maxPeriodBraveSearchPromotion.days : maxPeriodBraveSearchPromotion.minutes)
+    
+    if rightNow > nextShowDate {
+      return false
+    }
+      
+    return true
+  }
+  
+  // MARK: - Initialization
+
+  init(forTabType tabType: TabType, searchEngines: SearchEngines?) {
+    self.tabType = tabType
+    self.searchEngines = searchEngines
+  }
+  
+  func querySuggestClient() {
+    cancelPendingSuggestionsRequests()
+
+    let localSearchQuery = searchQuery.lowercased()
+    if localSearchQuery.isEmpty || searchEngines?.shouldShowSearchSuggestionsOptIn == true || localSearchQuery.looksLikeAURL() {
+      suggestions = []
+      delegate?.searchSuggestionDataSourceReloaded()
+      
+      return
+    }
+
+    suggestClient?.query(
+      localSearchQuery,
+      callback: { [weak self] suggestions, error in
+        guard let self = self else { return }
+
+        self.delegate?.searchSuggestionDataSourceReloaded()
+        if let error = error {
+          let isSuggestClientError = error.domain == SearchSuggestClientErrorDomain
+
+          switch error.code {
+          case NSURLErrorCancelled where error.domain == NSURLErrorDomain:
+            // Request was cancelled. Do nothing.
+            break
+          case SearchSuggestClientErrorInvalidEngine where isSuggestClientError:
+            // Engine does not support search suggestions. Do nothing.
+            break
+          case SearchSuggestClientErrorInvalidResponse where isSuggestClientError:
+            print("Error: Invalid search suggestion data")
```

Comment:
Should this use `Logger`? Same for print line below. I see they were just moved from other file so maybe not needed here.

Reply:
Fix at https://github.com/brave/brave-ios/pull/5432/commits/e13ac5a7e81aa9ce8df9852cad0f9b34774ca6f9